include/cuda/memory_resource

```diff
@@ -0,0 +1,318 @@
+{
+public:
+  static constexpr memory_kind kind = _Kind;
+  using context = _Context;
+
+  static constexpr std::size_t default_alignment = alignof(_CUDA_VSTD::max_align_t);
+
+  virtual ~memory_resource() = default;
+
+  /**
+   * @brief Returns the resource's execution context 
+   *
+   */
+  context get_context() const noexcept {
+      return do_get_context();
+  }
+
+  /**
+   * @brief Allocates storage of size at least `__bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `__alignment` if such
+   * alignment is supported.
+   *
+   * Storage may be accessed immediately within the execution context returned
+   * by `get_context()`, otherwise synchronization is required.
+   *
+   * @throws If storage of the requested size and alignment cannot be obtained.
+   *
+   * @param __bytes The size in bytes of the allocation
+   * @param __alignment The alignment of the allocation
+   * @return Pointer to the requested storage
+   */
+  void *allocate(std::size_t __bytes,
+                 std::size_t __alignment = default_alignment) {
+    return do_allocate(__bytes, __alignment);
+  }
+
+  /**
+   * @brief Deallocates the storage pointed to by `__p`.
+   *
+   * `__p` must have been returned by a prior call to `allocate(__bytes,
+   * __alignment)` on a `memory_resource` that compares equal to `*this`, and
+   * the storage it points to must not yet have been deallocated, otherwise
+   * behavior is undefined.
+   *
+   * @throws Nothing.
+   *
+   * @param __p Pointer to storage to be deallocated
+   * @param __bytes The size in bytes of the allocation. This must be equal to
+   * the value of `__bytes` that was specified to the `allocate` call that
+   * returned `__p`.
+   * @param __alignment The alignment of the allocation. This must be equal to
+   * the value of `__alignment` that was specified to the `allocate` call that
+   * returned `__p`.
+   */
+  void deallocate(void *__p, std::size_t __bytes,
+                  std::size_t __alignment = default_alignment) {
+    do_deallocate(__p, __bytes, __alignment);
+  }
+
+  /**
+   * @brief Compare this resource to another.
+   *
+   * Two resources compare equal if and only if memory allocated from one
+   * resource can be deallocated from the other and vice versa.
+   *
+   * @param __other The other resource to compare against
+   */
+  bool is_equal(memory_resource const& __other) const noexcept {
+    return do_is_equal(__other);
+  }
+
+private:
+  virtual void *do_allocate(std::size_t __bytes, std::size_t __alignment) = 0;
+
+  virtual void do_deallocate(void *__p, std::size_t __bytes,
+                             std::size_t __alignment) = 0;
+
+  // Default to identity comparison
+  virtual bool do_is_equal(memory_resource const &__other) const noexcept{
+      return this == &__other;
+  }
+
+  // Default to return default constructed Context object
+  virtual context do_get_context() const noexcept { return _Context{}; }
```

Comment:
This can't be defaulted to return a default constructed object if `_Context`'s default ctor is deleted. 
```suggestion
  virtual context do_get_context() const noexcept = 0;
```